### DISCUSSION TITLE
error while requesting for random number generator on chainlink vhf

### DISCUSSION THREAD

**Discussion by johan784** [Cyfrin/foundry-full-course-cu#2910] [Q&A] [answered]
I've created a VRF subscription and received a subscription ID as well, however, the id has a very long range of numbers in it and I'm unable to deploy the vrf contract which accepts only a uint64 as an input for the subscription id

however in the chainlink tutorial provided, the subscription ID the instructor received was only 4 digits long

is there something wrong with my id??

**Comment by KMean** [comment]
Are you sure you're using the new VRF 2.5? the old (v2) version is using a uint64 for subscriptionId while the new one uint256.

https://docs.chain.link/vrf/v2-5/migration-from-v2

**Reply by johan784** [reply]
hey I did switch to v2.5 but now I seem to face a different issue, let me paste it down here 

![image](https://github.com/user-attachments/assets/6b89653a-e107-443f-a090-0f87e88e7197)

I have successfully funded my subsciption with around 10 Link , however when I interact with the "request random words " button of the vrf contract ( I entered "false" since I wanted to make my request by spending link and not native eth tokens) I am getting a revert error, could you help me out?

**Reply by KMean** [reply]
Have you added the consumer address to your subscription? 
You need to copy the address of your deployed contract (if using remix ensure you're deploying it to sepolia through injected provider) and add it as a consumer to your subscritpion. wait for the add consumer transaction to go through. Make sure your subscription is funded (which you said you already did).  Try requesting the numbers again.

![Screenshot 2024-11-23 121627](https://github.com/user-attachments/assets/c9e748ff-c4b7-45ab-87c2-d79f2ebaab66)

![Screenshot 2024-11-23 121451](https://github.com/user-attachments/assets/bdce20bc-fc04-41bd-acfb-a8eaa59087cd)

**Reply by johan784** [reply]
I'm facing a gas estimation error, any way to fix it?

**Reply by KMean** [reply, marked as answer]
can you show me a screenshot of your funded subscription? like the one I've posted? Maybe you haven't deployed to sepolia?

Ensure you're using injected provider when deploying and that your metamask is on sepolia (chainid 11155111).
Enter your subid in the field and click deploy. 
Metamask pops up and ask you to confirm (you need to have sepolia eth). 
confirm
do the steps I've mentioned before:
copy the address of the deployed contract.
add it as a consumer to your subscription.
request random numbers. (metamask again pops up asking you to confirm).
![Screenshot 2024-11-23 123446](https://github.com/user-attachments/assets/5caacf20-3af6-4a8d-b38c-10235da7689f)

**Reply by johan784** [reply]
![image](https://github.com/user-attachments/assets/446811ce-dba9-4315-86be-2882be905855)

**Reply by johan784** [reply]
ive been trying to deploy the contract to add a consumer on injected metamask but Ive been receiving gas estimation errors even if I set a custom gas limit

**Reply by KMean** [reply]
do you have sepolia eth to pay for the transaction?

**Reply by johan784** [reply]
I have 0.19 sepoliaeth

**Comment by johan784** [comment]
![image](https://github.com/user-attachments/assets/3360c7bb-2437-4d47-87c4-33b00170a8f7)

it worked! how long do you presume it takes for the transaction to go through?

**Reply by johan784** [reply]
![image](https://github.com/user-attachments/assets/ac028dec-7a3b-40e6-87b7-edffa1339fbe)


would this mean that my status is still pending and that may be the reason why ive received an empty array?

**Reply by EngrPips** [reply]
Hello, Yes it is still pending and yet to be processed.How many Links did you fund your subscription with?

**Reply by KMean** [reply]
> ![image](https://private-user-images.githubusercontent.com/175939955/389194737-3360c7bb-2437-4d47-87c4-33b00170a8f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI0NDQ1NzYsIm5iZiI6MTczMjQ0NDI3NiwicGF0aCI6Ii8xNzU5Mzk5NTUvMzg5MTk0NzM3LTMzNjBjN2JiLTI0MzctNGQ0Ny04N2M0LTMzYjAwMTcwYThmNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNFQxMDMxMTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ODk5OTlkNDU0OWNkNzczODNiY2FhNTE1MDg5NGQxMDgwNGJhYjdiYWYwZThkZjkyYTg4Y2RkNWY0NjhiYTcyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.PeS83nmfjnNbcya1bqzyqHbiqdZVAh2gYrGCmdJURdE)
> 
> it worked! how long do you presume it takes for the transaction to go through?

That actually mean that probably your transaction will never go trough due to low balance on your subscription. Calculating the fund needed to keep your requests going is sometime a little bit challenging and you can refer to this to have a better understanding:

https://docs.chain.link/vrf/v2-5/billing#subscription-cost-examples

Why is that happening?  Probably because sepolia is really busy at that moment and the gas price is very high which means that eventually if the gas price is lowering before 24 hours your transaction will be picked up otherwise will be canceled. You can read more about it in this discussion on the chainlink discord:

https://discord.com/channels/592041321326182401/722867054197080136/1288818177983250474

Said that increasing your funds to be more then the max cost will make your transaction go through straight away and that doesnt mean you will pay all that links (29.70 in your case) to get the random numbers but the funds are needed to protect from gas price spikes.

I think needing 30-35 LINK (at currently 17$ each) just to request random numbers is mental but remember we are talking about ethereum where a simple send transaction can cost you more then 40 bucks. In a real world situation I would never use an L1 for an application that frquently request random numbers.

**Reply by LordAdhaar** [reply]
There must be solution for this on polygon.

**Reply by EngrPips** [reply]
Interesting.

**Reply by johan784** [reply]
makes total sense, thanks!

**Reply by KMean** [reply]
> There must be solution for this on polygon.

Absolutely... using for example Polygon Amoy to do the test you don't need to fund that much your subscription. Just change the vrfCoordinator and the keyhash following https://docs.chain.link/vrf/v2-5/supported-networks. Very cheap and faster.
![Screenshot 2024-11-25 192142](https://github.com/user-attachments/assets/1ef51b74-dfa9-4c8b-bdd1-231e10f8ffba)

**Reply by EngrPips** [reply]
Awesome, Great heads-up

**Reply by klovistore2** [reply]
Hello I'm facing the exact same problem, both on SEPOLIA and POLYGON MAINNET, did you solve it ? 

If yes I'm looking for a very simple code that works on VRF2.5.

I tried every possible values for   uint32 callbackGasLimit  

if I use 200 000 : on polygon mainnet I have : Failed: gas limit set too low, 

if I put 3 000 000 : reason: 'execution reverted', code: 'UNPREDICTABLE_GAS_LIMIT',

Thank you for any help

**Reply by EngrPips** [reply]
> Hello I'm facing the exact same problem, both on SEPOLIA and POLYGON MAINNET, did you solve it ?
> 
> If yes I'm looking for a very simple code that works on VRF2.5.
> 
> I tried every possible values for uint32 callbackGasLimit
> 
> if I use 200 000 : on polygon mainnet I have : Failed: gas limit set too low,
> 
> if I put 3 000 000 : reason: 'execution reverted', code: 'UNPREDICTABLE_GAS_LIMIT',
> 
> Thank you for any help

Have you seen the `lottery` course on Cyfrin Updraft? I believe Patrick teaches how to implement `VRF2.5`.
